### ISSUE TITLE
Motis' Gradients Values

### ISSUE THREAD

**Issue by panizbehboudian** [xingchenwan/nasbowl#1]
Hi there!

I have read you awesome work and decided to implement the motif extraction on my domain. I noticed that in the paper you have mentioned that `We finally rank the features by the numerical values of the
derivatives, and present the top and bottom quantiles of the features as “best motifs” and “worst
motifs” respectively `  and  also `motifs with most positive and negative derivatives (hence the most and least desirable)`. However in the interpreter.py code where you want to extract good and bad motifs you do this
`````python

        up = np.percentile(self.grad_, 100. - thres)
        lo = np.percentile(self.grad_, thres)

        # Identify the good and bad features
        self.bad_idx = np.argwhere(self.grad_ >= up).flatten()
        self.good_idx = np.argwhere(self.grad_ <= lo).flatten()
``````
it seems that in the code you are doing the opposite: considering the smallest gradients corresponding to the best motifs and vice versa. Can you please explain more about this?

Also while I was looking at the motifs' gradients on my data, I notice a large portion of the motifs having 0 gradients. Can you help me interpreting this skewed distribution? Whether I need to check something specific for this?

Thank you in advance for your kind help,

Paniz

**Comment by xingchenwan** [comment]
Hi Paniz thanks for your interest.
On your first question: in paper for convenience of notation we defined the derivative w.r.t **accuracy**. Here in codes the derivative is computed w.r.t **error** (1 - accuracy) and hence the flipping of the sign.

On your second question: this is expected as many of the subgraphs found will not have influence on the accuracy of an architecture and often the performance of architecture is largely influenced by a few recurring subgraphs only. Derivative around 0 essentially means the motif has little influence on the architecture performance (as far as the GPWL surrogate is accurate).

Note this is different from very negative gradients, as gradient with large magnitude on either the positive and negative direction suggests importance.

Hope that helps.

Xingchen

**Comment by TomKan0909** [comment]
Hi @xingchenwan !

Hope you're having a great day! I have a question regarding the choice of how the gradient values are being selected for each motif/feature. The code I'm referring to is in bayesopt/interpreter.py. 

Correct me if I'm wrong but from what I understand the gradient value picked for each motif/feature is the gradient value that occurs at index 1 for it's corresponding gradient tensor in self.grads. Having stated that, I was wondering why is index 1 chosen as the gradient value for the motifs. 
![Capture93](https://user-images.githubusercontent.com/43018123/126010813-b214989b-98c1-4e3b-94a7-4d90ca530ee6.PNG)

Thanks in advance!
Tom

**Comment by xingchenwan** [comment]
@TomKan0909  Hi Tom! Would you mind letting me check? It's a busy week for me and the code was written a while ago, but rest assured I will get back.

**Comment by TomKan0909** [comment]
@xingchenwan no worries and thanks a lot!

**Comment by xingchenwan** [comment]
@TomKan0909 

Hi Tom,

Many thanks for your patience.

The averaged gradient is computed according to Eq 3.1 in the paper, which averages gradient over all instances where a feature has occurred more than 0 times (i.e. whether a feature exists).

The [1] index is chosen as a quick trick to approximate Eq 3.1, as given the size of the graph (recall that DAGs in these NAS datasets are quite small), we find that a feature (except for those h=0, which are the node types and not motif patterns which we are interested in) is usually either present 1 time or not present at all: \phi^j(G) is almost always binary -- either 0 or 1. Even if there are few instances of \phi^j(G) > 1, the weight of these instances (noting the p(\phi^j(g)) term) is very small. Given that, we don't actually need to compute the integral (or more precisely a summation, since \phi^j(G) is an integer which denotes the number of times feature \phi^j occurs in graph G) in Eq 3.1, and just taking the second element of the grad vector will suffice.

This is indeed a small trick for our problem, and if you are using the code repo for other purposes where your graph properties are different, I recommend fully implementing Eq 3.1.  It is very straightforward as a weighted summation, and I might release a new version which properly implement Eq 3.1 on a future date. 

Hope that helps,
Xingchen

**Comment by TomKan0909** [comment]
Hi @xingchenwan 

Thanks for the reply. For Eq 3.1 are you referring to this?
![Capture97](https://user-images.githubusercontent.com/43018123/127682354-fd57b356-a2df-41dc-ba36-3e3b63b34522.PNG)

As for the weighted summation. Correct me if I am wrong but I believe the code has an existing way of calculating it. I set average_across_occurrences flag to True when calculating the gradients in bayesopt/interpreter.py which appears to be doing a weighted summation of all the gradient values for a given motif. 
![Capture98](https://user-images.githubusercontent.com/43018123/127682612-a593dab6-f139-4f8e-a8d2-b4a452585ff7.PNG)

Once again, thanks for your reply and guidance!

Cheers,
Tom

**Comment by xingchenwan** [comment]
Hi @TomKan0909,
Sorry for the rather late reply! Yes that is the equation I am referring to (sorry it should be Eq 3.3 instead of 3.1).

Now, consider if we don't do any averaging at all, and we compute the derivatives w.r.t each feature \phi for all the training samples, at all possible values (note that even if, say, for graph G, \phi occurs 1 time, we can still compute the grad at 0, 1,..., D for that graph. However, as you'd expect, these gradients except for the one at 1 will not be as accurate as they involve extrapolations. In general, taking grad, which is a first-order operation, amplifies noise even more and making them even less reliable than the predictions), we would've got a matrix with size N x D, where N is the number of training samples and D is the maximum number of occurrences (we can manually set D to some value like 7, as we don't expect a single graph to contain 7 such features).

The most proper way to implement the equation is 1) for each graph, we check how many times \phi has actually occurred, and cache the grad corresponds to that value (e.g. if \phi occurs 2 times for i-th graph, we should cache the [i, 2]-th element of the said matrix) 2) for all training graphs, we compute the weights (how many graphs contain 1 \phi, how many 2 \phi, ...) and 3) do the averaging.

However, since we observe that in most cases, \phi is  either 1 and 0, what we do is 1) average on the N dimension (which is what average_across_occurrences does) to obtain a D-dimensional vector and 2) take the second element of the vector (which corresponds to the grad where the feature exists).

So, in short, average_across_occurrences does the weighted summation across all training samples, and since we know most features either exist 0 or 1 time, the [:, 1] approximates the grad where the feature exists.

The proper way to implement Eq 3.3 is very straightforward, and as I said, I would possibly push a new version very soon.

**Comment by xingchenwan** [comment]
Closing this issue for now -- please feel free re-opening it if you have further questions!
